apiserver/facades/client/application/application.go

```diff
@@ -749,6 +797,39 @@ func deployApplication(
 	return errors.Trace(err)
 }
 
+func convertCharmOrigin(origin *params.CharmOrigin) (corecharm.Origin, error) {
+	switch {
+	case origin == nil || origin.Source == "" || origin.Source == "unknown":
+		return corecharm.Origin{Source: corecharm.Unknown}, nil
+	case origin.Source == "local":
+		return corecharm.Origin{Source: corecharm.Local}, nil
+	case origin.Source == "charm-store":
+		return corecharm.Origin{Source: corecharm.CharmStore}, nil
+	}
+
+	// Charmhub code can not tell us about the exact charm we want to install.
+	if origin.Revision == nil && origin.Channel == nil {
```

Comment:
A revision can never be nil unless it's unknown.  All 3 types must have a revision.

Reply:
Fair